File: DSA/DSA04009 LŨY THỪA MA TRẬN 1.cpp
#include <bits/stdc++.h>
using namespace std;

int MOD = 1e9+7;

void multiply(vector<vector<long long>> &A, vector<vector<long long>> &B, int n)
{
    vector<vector<long long>> C(n, vector<long long>(n, 0));

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            for (int k = 0; k < n; k++)
            {
                C[i][j] += (A[i][k]*B[k][j]);
            }
        }
    }
    
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            A[i][j] = C[i][j]%MOD;
        }
    }
}

vector<vector<long long>> setup(vector<vector<long long>> &A, int n, int k)
{
    vector<vector<long long>> ans(n, vector<long long>(n, 0));
    for (int i = 0; i < n; i++)
    {
        ans[i][i] = 1;
    }
    
    while (k > 0)
    {
        if (k&1)
        {
            multiply(ans, A, n);
        }
        multiply(A, A, n);
        k >>= 1;
    }
    return ans;
}

void submit()
{
    int n, k;
    cin >> n >> k;
    
    vector<vector<long long>> M(n, vector<long long>(n));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> M[i][j];
        }
    }
    
    vector<vector<long long>> ans = setup(M, n, k);
    
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
} 

File: DSA_python/DSA04009 LŨY THỪA MA TRẬN 1.py
def multiply(A, B, n):
    C = [[0]*n for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += (A[i][k]*B[k][j])
            C[i][j] %= MOD
    return C

def setup(A, n, k):
    ans = [[0]*n for _ in range(n)]
    for i in range(n):
        ans[i][i] = 1
    
    while k > 0:
        if k&1:
            ans = multiply(ans, A, n)
        A = multiply(A, A, n)
        k >>= 1

    return ans


def submit():
    n, k = map(int, input().split())

    M = []
    for i in range(n):
        A = list(map(int, input().split()))
        M.append(A[:])
    
    ans = setup(M, n, k)

    for i in ans:
        print(" ".join(map(str, i)))


MOD = 1000000007

test = int(input())
for i in range(test):
    submit()